Add sensor data to harvest list

mhauru opened the pull request:
There's still plenty I want to do here, but probably won't get to it tomorrow or Friday, and since this is already an improvement, may as well merge.

This adds averages of T&RH data to the harvest_list table from the relevant sensors in the relevant time periods. This is done through a massive SQL query that is currently just written as a string. That's because the old way of doing things using pandas would have been very slow, and I needed to learn a bunch of SQL to do the heavy lifting on the postgres server. That massive query needs to be converted into SQLAlchemy commands. Much of those can then be reused to implement batch_list and batch_details more efficiently and elegantly. Some performance optimisation could also be done for the query. It's currently bearable, but might get quite slow once more crop data comes in.

Also fix a bug introduced in batch_list by the previous PR on this topic.

mhauru commented:
I've now converted the SQL query into SQLAlchemy and used it to also implement a first version of Zack's parallel axes plot. Ready for review when you have a moment @nbarlowATI. I'll open issues for some tweaks I have in mind that I didn't yet get to.

nbarlowATI commented on core/queries.py:
```diff
@@ -0,0 +1,276 @@
+        .outerjoin(grow_trh, grow_trh.c.batch_id == BatchClass.id)
+        .outerjoin(propagate_trh, propagate_trh.c.batch_id == BatchClass.id)
+    )
+    return query
```
That's quite a query! :).   I like the idea of having this core module containing queries though - feels like we can migrate some other stuff there at some point...

mhauru replied:
Agreed:
https://github.com/alan-turing-institute/CROP/issues/371
https://github.com/alan-turing-institute/CROP/issues/374

nbarlowATI reviewed (commented):
This looks great!   Both the querying and the new parallel axes page.  
A couple of comments:
* For me, on a non-4k monitor, the axis labels overlap quite a bit.... Maybe we just have too many axes?  Maybe we don't need all 3 of temp, RH, and VPD?   Or maybe we could choose between yield, waste defect, waste disease, overproduction via a dropdown?  (Not sure if any of these are good ideas, but we could perhaps discuss them and more at next meeting?)
* For certain choices of crop, e.g. "broccoli (micro)" I get a crash - perhaps if the query returns no results for the specified timespan?   The traceback is:
```
File "/Users/nbarlow/CROP/webapp/app/crops/routes.py", line 584, in parallel_axes
df[column_name] = df[column_name].dt.total_seconds() / 3600 / 24
File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/generic.py", line 5487, in __getattr__
return object.__getattribute__(self, name)
File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/accessor.py", line 181, in __get__
accessor_obj = self._accessor(obj)
File "/Users/nbarlow/anaconda3/envs/cropenv/lib/python3.8/site-packages/pandas/core/indexes/accessors.py", line 506, in __new__
raise AttributeError("Can only use .dt accessor with datetimelike values")
```

mhauru commented:
I fixed the crash and added checkboxes for choosing which columns to display. By default all are displayed which I agree is too much, can adjust that based on e.g. feedback from Jakob.

nbarlowATI reviewed (approved):
Looks good to me, and works well!   Lets see if we end up with a smaller list of variables that are useful, but in the meantime this checkbox system is an excellent way of narrowing it down.
